File: day10/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

// The manual describes one machine per line. Each line contains a single indicator light diagram in [square brackets], one or more button wiring schematics in (parentheses), and joltage requirements in {curly braces}.

type machine struct {
	numLights    uint   // integer number of lights
	lights       uint   // 0 bit represents that light should be off, and 1 bit represents that light should be on
	buttons      []uint // use the bits to indicate which lights are controlled by the button
	requirements []uint // num joltage requirements == num lights
}

func printMachine(m *machine) {
	fmt.Printf("Machine:\n")
	fmt.Printf("%d, %d\n", m.numLights, m.lights)
	fmt.Printf("%+v\n", m.buttons)
	fmt.Printf("%+v\n", m.requirements)
}

func assert(expr bool, message ...string) {
	if !expr {
		if len(message) > 0 {
			log.Printf("[FATAL] %s\n", message[0])
		}
		panic(expr)
	}
}

func powTwo(exp int) int {
	var result int = 2
	for range exp {
		result = result << 1
	}

	return result
}

func populateLights(m *machine, lightStr string) {
	lightStr = lightStr[1 : len(lightStr)-1]
	m.numLights = uint(len(lightStr))
	m.lights = uint(0)

	for i, b := range lightStr {
		if b == '#' {
			m.lights += (1 << i)
		}
	}
}

func populateButtons(m *machine, buttonStrs []string) {
	assert(m.numLights != 0)
	m.buttons = make([]uint, 0, 32)

	for btnIdx, buttonStr := range buttonStrs {
		m.buttons = append(m.buttons, uint(0))

		buttonStr = buttonStr[1 : len(buttonStr)-1]
		for digit := range strings.SplitSeq(buttonStr, ",") {
			d, err := strconv.Atoi(digit)
			if err != nil {
				log.Fatalf("failed to convert digit %s\n", digit)
			}
			m.buttons[btnIdx] += (1 << d)
		}
	}
}

func populateRequirements(m *machine, reqStr string) {
	m.requirements = make([]uint, 0, m.numLights)

	reqStr = reqStr[1 : len(reqStr)-1]
	for digit := range strings.SplitSeq(reqStr, ",") {
		d, err := strconv.Atoi(digit)
		if err != nil {
			log.Fatalf("failed to convert digit %s to requirement\n", digit)
		}

		m.requirements = append(m.requirements, uint(d))
	}
	assert(len(m.requirements) == int(m.numLights), fmt.Sprintf("req count %d doesn't equal num lights %d\n", len(m.requirements), int(m.numLights)))
}

func readMachine(line string) machine {
	m := machine{}

	// input format
	// [.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
	// now that I know the minimum pushes to get the lights, I can push each button an even number of times to get to the joltage

	fields := strings.Fields(line)
	assert(len(fields) > 0)

	populateLights(&m, fields[0])

	populateButtons(&m, fields[1:len(fields)-1])

	populateRequirements(&m, fields[len(fields)-1])
	return m
}

func toggle(m *machine, lights uint, buttonIdx int) uint {
	return lights ^ m.buttons[buttonIdx]
}

func minButtonPushes(m *machine) int {
	return minButtonPushesRecursive(m, 0, 0)
}

func min(x int, y int) int {
	return -max(-x, -y)
}

func minButtonPushesRecursive(m *machine, current uint, startIdx int) int {
	if current == m.lights {
		return 0
	}

	nButtons := len(m.buttons)
	if startIdx == nButtons {
		return math.MaxInt
	}

	// compare toggling current button with skipping
	return min(1+minButtonPushesRecursive(m, toggle(m, current, startIdx), startIdx+1), minButtonPushesRecursive(m, current, startIdx+1))
}

func joltageEquals(req, current []uint) int {
	if len(req) != len(current) {
		panic("Lengths differ")
	}

	equal := 0
	for i := range req {
		if req[i] < current[i] {
			return -1
		}
		if req[i] > current[i] {
			equal = 1
		}
	}
	return equal
}

func minButtonPushesP2(m *machine) int {
	current := make([]uint, len(m.requirements))
	for i := range current {
		current[i] = 0
	}

	return minButtonPushesRecursiveP2(m, current, 0, 0)
}

func pressButton(m *machine, current []uint, idx int) {
	for i := range 64 {
		if (m.buttons[idx]>>i)&1 == 1 {
			current[i] += 1
		}
	}
}

func unPressButton(m *machine, current []uint, idx int) {
	for i := range 64 {
		if (m.buttons[idx]>>i)&1 == 1 {
			current[i] -= 1
		}
	}
}

func minButtonPushesRecursiveP2(m *machine, current []uint, startIdx int, depth int) int {
	cmp := joltageEquals(m.requirements, current)
	if cmp == 0 {
		return 0
	}
	if cmp == -1 {
		return math.MaxInt
	}

	nButtons := len(m.buttons)
	if startIdx == nButtons {
		return math.MaxInt
	}

	pressButton(m, current, startIdx)
	minForPressing := 1 + minButtonPushesRecursiveP2(m, current, startIdx, depth+1)
	unPressButton(m, current, startIdx)
	minForMoving := minButtonPushesRecursiveP2(m, current, startIdx+1, depth+1)
	fmt.Println(min(minForPressing, minForMoving))

	return min(minForPressing, minForMoving)

}

const INPUT_FILENAME = "input.txt"

func p1() {
	f, err := os.Open(INPUT_FILENAME)
	if err != nil {
		log.Fatalf("Could not open file. %s\n", err.Error())
	}

	var total int = 0

	var scanner *bufio.Scanner = bufio.NewScanner(f)
	for scanner.Scan() {
		var m machine = readMachine(scanner.Text())
		total += minButtonPushes(&m)
	}

	fmt.Printf("Part one: %d\n", total)
}

func p2() {
	f, err := os.Open(INPUT_FILENAME)
	if err != nil {
		log.Fatalf("Could not open file. %s\n", err.Error())
	}

	var total int = 0

	var scanner *bufio.Scanner = bufio.NewScanner(f)
	for scanner.Scan() {
		var m machine = readMachine(scanner.Text())
		total += minButtonPushesP2(&m)
	}

	fmt.Printf("Part two: %d\n", total)
}

func main() {
	p1()
	// p2()
}

File: day10/c/main.c
#include <assert.h>
#include <limits.h>
#include <errno.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../logging.h"
#include "../../slice.h"

#define INPUT_FILENAME "input.txt"
#define LINE_SIZE 1024

typedef struct {
  uint64_t numLights;
  uint64_t lights;
  slice *buttons;
  slice *requirements;
} machine;

machine *new_machine() {
  machine *m = malloc(sizeof(machine));
  m->numLights = 0;
  m->lights = 0;
  m->buttons = slice_new_empty(0, 64, sizeof(uint64_t));
  m->requirements = NULL;

  return m;
}

void print_machine(machine *m) {
	printf("Machine:\n");
	printf("%lld, %lld\n", m->numLights, m->lights);
  printf("[");
  for (size_t i = 0; i < slice_length(m->buttons); i++) {
	printf("%lld ", *(uint64_t*)slice_get(i, m->buttons));
  }
  printf("]\n");
  printf("[");
  for (size_t i = 0; i < slice_length(m->requirements); i++) {
	  printf("%lld ", *(uint64_t*)slice_get(i, m->requirements));
  }
  printf("]\n");
}

void free_machine(machine *m) {
  slice_free(m->buttons);
  slice_free(m->requirements);
}

void populateLights(machine *m, char *word) {
  size_t len = strlen(word);
  if (word[len - 1] == '\n') {
    word[len-1] = '\0';
    len--;
  }

  m->numLights = (uint64_t) len - 2;
  m->lights = (uint64_t) 0;

  for (size_t i = 1; i < len - 1; i++) {
    if (word[i] == '#') {
      m->lights |= (1<<(i-1));
    }
  }
}

void addButton(machine *m, char *word) {
  size_t len = strlen(word);
  if (word[len - 1] == '\n') {
    word[len - 1] = '\0';
    word[len - 2] = '\0';
    len -= 2;
  } else {
    word[len - 1] = '\0';
    len -= 1;
  }
  word++;
  
  uint64_t btn = 0;
  char *lasts;
  for (char *digit = strtok_r(word, ",", &lasts); digit; digit = strtok_r(NULL, ",", &lasts)) {
    char *endPtr;
    uint64_t d = strtol(digit, &endPtr, 10);
    if (*endPtr != '\0') {
      log_fatal("Failed to convert digit %s\n", digit);
    }

    btn |= (1 << d);
  }

  slice_append((void *) &btn, m->buttons);
}

void populateRequirements(machine *m, char *word) {
  size_t len = strlen(word);
  if (word[len - 1] == '\n') {
    word[len - 1] = '\0';
    word[len - 2] = '\0';
    len -= 2;
  } else {
    word[len - 1] = '\0';
    len -= 1;
  }
  word++;

  m->requirements = slice_new_empty(0, m->numLights, sizeof(uint64_t));
  
  char *lasts;
  for (char *digit = strtok_r(word, ",", &lasts); digit; digit = strtok_r(NULL, ",", &lasts)) {
    char *endPtr;
    uint64_t d = (uint64_t) strtol(digit, &endPtr, 10);
    if (*endPtr != '\0') {
      log_fatal("Failed to convert req digit %s\n", digit);
    }

    slice_append((void *) &d, m->requirements);
  }
  assert(slice_length(m->requirements) == (size_t) m->numLights);
}

machine *readMachine(char *line) {
  machine *m = new_machine();

  char *lasts;
  for (char *word = strtok_r(line, " ", &lasts); word; word = strtok_r(NULL, " ", &lasts)) { 
    switch (word[0]) {
      case '[':
        populateLights(m, word);
        break;
      case '(':
        addButton(m, word);
        break;
      case '{':
        populateRequirements(m, word);
        break;
    }
  }

  return m;
}

long min(long x, long y) {
  return x * (x < y) + y * (y <= x);
}

uint64_t toggle(machine *m, uint64_t lights, size_t buttonIdx) {
  return lights ^ *(uint64_t*) slice_get(buttonIdx, m->buttons);
}

long minButtonPushesRecursive(machine *m, uint64_t current, size_t startIdx) {
  if (current == m->lights) {
    return 0; 
  }

  if (startIdx == slice_length(m->buttons)) {
    return LONG_MAX - 1;
  }

  return min(1 + minButtonPushesRecursive(m, toggle(m, current, startIdx), startIdx + 1), minButtonPushesRecursive(m, current, startIdx + 1));
}

long minButtonPushes(machine *m) {
  return minButtonPushesRecursive(m, 0, 0);
}


void p1() {
  FILE *f = fopen(INPUT_FILENAME, "r");
  if (f == NULL) {
    log_fatal("Could not open file. %s", strerror(errno));
  }

  long total = 0;
  char line[LINE_SIZE];
  while (fgets(line, LINE_SIZE, f) != NULL) {
    machine *m = readMachine(line);
    total += minButtonPushes(m);

  }
  if (ferror(f)) {
    log_fatal("Error reading lines from file. %s", strerror(errno));
  }
  fclose(f);

  printf("Part one: %ld\n", total);
}

int main() {
  p1();
}
